### PR TITLE
Create Rest Routing Bundle

### PR DESCRIPTION
**Pull request by alexander-schranz**
This moves the routing loader from the FOSRestBundle into an own Bundle as it was removed in 3.0. The classes the old classes are aliased to allow a seamless update.

~~Requires: FriendsOfSymfony/FOSRestBundle#2168~~

### CODE REVIEW

File: DependencyInjection/FOSRestRoutingExtension.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+        $container->getDefinition('fos_rest.routing.loader.yaml_collection')->replaceArgument(4, $config['routing_loader']['default_format']);
+        $container->getDefinition('fos_rest.routing.loader.xml_collection')->replaceArgument(4, $config['routing_loader']['default_format']);
+
+        $container->getDefinition('fos_rest.routing.loader.yaml_collection')->replaceArgument(2, $config['routing_loader']['include_format']);
+        $container->getDefinition('fos_rest.routing.loader.xml_collection')->replaceArgument(2, $config['routing_loader']['include_format']);
+        $container->getDefinition('fos_rest.routing.loader.reader.action')->replaceArgument(3, $config['routing_loader']['include_format']);
+        $container->getDefinition('fos_rest.routing.loader.reader.action')->replaceArgument(5, $config['routing_loader']['prefix_methods']);
+
+        $container->setParameter('fos_rest.formats', [
```
Comment (alexander-schranz) [inline DependencyInjection/FOSRestRoutingExtension.php:51]: Need to be replaced with https://github.com/FriendsOfSymfony/FOSRestBundle/pull/2168

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        "files": ["aliases.php"],
+        "exclude-from-classmap": [
+            "Resources/",
+            "Tests/"
+        ]
+    },
+    "require": {
+        "php": "^7.2",
+        "psr/log": "^1.0",
+        "friendsofsymfony/rest-bundle": "^3.0@dev",
```
Comment (alexander-schranz) [inline composer.json:20]: Need to be removed when fos rest bundle is removed.

**Comment by Tobion** [comment]
I'd prefer to not put this into the FOS namespace or organisation. Otherwise we have the same problem as before: who is gonna maintain this?
Do you really need these features and cannot use explicit routing instead? For sulu? If that is the case, I'd suggest to keep it in your personal or sulu org so it's clear who is responsible for it.
Also this is a chance to actually do some BC breaks to that would be required to get the auto-routing up-to-date with current symfony best practices. See also the issues I referenced in https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1954

**Comment by alexander-schranz** [comment]
> Otherwise we have the same problem as before: who is gonna maintain this:

I'm ok with letting it in this organization and don't move it to the FOS organization, I see no difference in which organisation it is on github when you think nobody else of the FOS organization team will do anything with it. And if it is so it could still be later moved to the friendsofsymfony organization.

> Do you really need these features and cannot use explicit routing instead? For sulu?

> Also this is a chance to actually do some BC breaks

Thats basically the whole thing why we building it to have as less BC Breaks as possible on the sulu side and make the update seamless as possible. So I don't think we will change here much, else as you suggestion we could define every route ourselves over annotations or yaml.

**Comment by lsmith77** [comment]
@Tobion I know you don’t see much point in this Bundle but from much having migrated away from this functionality in a code base, I can say that using the standard routing I ran into several trivial typos which led to an inconsistent API .. especially around pluralization with collections but also other aspects. Not to mention a ton of extra config code to maintain.

By splitting the Bundle imho we make maintenance a lot easier and if we have a healthy project using this functionality I think the near/mid-term maintenance should be covered.

Furthermore it will ease the migration path for people wanting to use newer version of FOSRestBundle that rely on the routing generation.

So I would welcome such a Bundle under the condition that Sulu project team takes responsiblity for maintaining it.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
+FOSRoutingRestBundle
+=============
+
+This bundle provides the automatic route generation for to FOSRestBundle 3.0
```
Comment (xabbuh) [inline README.md:4]: for to FOSRestBundle 3.0.

File: Routing/ClassResourceInterface.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\RestRoutingBundle\Routing;
+
+use FOS\RestBundle\Routing\ClassResourceInterface as OldClassResourceInterface;
+
+if (!class_exists(OldClassResourceInterface::class)) {
+    class_alias(ClassResourceInterface::class, OldClassResourceInterface::class);
```
Comment (xabbuh) [inline Routing/ClassResourceInterface.php:17]: What ist the reason for this?

File: composer.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        },
+        "exclude-from-classmap": [
+            "Resources/",
+            "Tests/"
+        ]
+    },
+    "require": {
+        "php": "^7.2",
+        "doctrine/inflector": "^1.0",
+        "friendsofsymfony/rest-bundle": "^3.0@dev",
```
Comment (xabbuh) [inline composer.json:19]: I would change this to `^3.0` and add `minimum-stability` to `dev` instead. So you will not have to change this in the future.

File: DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        $treeBuilder = new TreeBuilder('fos_rest_routing');
+
+        $rootNode = $treeBuilder->getRootNode();
+
+        $rootNode
+            ->children()
+                ->arrayNode('routing_loader')
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->booleanNode('parse_controller_name')
```
Comment (xabbuh) [inline DependencyInjection/Configuration.php:47]: I would remove this option and have only the code that would be executed when it is `false`.

**Comment by danrot** [comment]
> @Tobion I know you don’t see much point in this Bundle but from much having migrated away from this functionality in a code base, I can say that using the standard routing I ran into several trivial typos which led to an inconsistent API .. especially around pluralization with collections but also other aspects. Not to mention a ton of extra config code to maintain.

I think I've also already made my point, that this was the best part about the `FOSRestBundle`. I still think that the rest of the features we use are not worth risking struggles at every major Symfony update, as we have them currently.

> By splitting the Bundle imho we make maintenance a lot easier and if we have a healthy project using this functionality I think the near/mid-term maintenance should be covered.
> 
> Furthermore it will ease the migration path for people wanting to use newer version of FOSRestBundle that rely on the routing generation.
> 
> So I would welcome such a Bundle under the condition that Sulu project team takes responsiblity for maintaining it.

When only the Sulu team is going to work on that, it might really make more sense to keep it in one of our namespaces IMO.

File: DependencyInjection/Configuration.php (hunk 6 above)
Comment (alexander-schranz) [reply DependencyInjection/Configuration.php]: Make sense, good catch 👍

File: Routing/ClassResourceInterface.php (hunk 4 above)
Comment (alexander-schranz) [reply Routing/ClassResourceInterface.php:17]: We want to have as less BC Breaks as possible so we still support the old used classes and namespace.

Comment (xabbuh) [reply Routing/ClassResourceInterface.php:17]: Makes sense, but is there a reason not to do the same with the annotations for example?

Comment (alexander-schranz) [reply Routing/ClassResourceInterface.php:17]: @xabbuh seems like that got lost accidently need to fix it, thank you for your review!

**Comment by lsmith77** [comment]
@danrot I am fine if you keep it in your namespace. I would appreciate if you however make sure that you make at least one fully BC stable release before making any BC changes in a new major version, so that we can announce it as a simple upgrade path.

**Comment by danrot** [comment]
@lsmith77 I think that was the plan anyway :slightly_smiling_face: We want to keep BC as good as possible, so that we don't have to change as less as possible in our codebase. And I think we don't plan any huge BC breaks here, but would like to keep it working as it currently is.

@alexander-schranz Do you have the same view on this?

**Comment by alexander-schranz** [comment]
@lsmith77 @danrot at current state it looks like we have no bc break from the code side. The developer will just need to install and configure the bundle. We are still supporting the old classes using class_aliases and using a compilerpass to read some configuration out of the FOSRestBundle if it exists.

File: composer.json [hunk 7]
```diff
@@ -0,0 +1,52 @@
+{
+    "name": "friendsofsymfony/rest-routing-bundle",
+    "type": "symfony-bundle",
+    "description": "This Bundle automatic route registration for the Controllers",
```
Comment (xabbuh) [inline composer.json:4]: "This bundle provides [...]"?

File: Tests/Fixtures/Controller/OldTypeHintedController.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\RestRoutingBundle\Tests\Fixtures\Controller;
+
+use FOS\RestBundle\Controller\Annotations\NamePrefix;
+use FOS\RestBundle\Controller\Annotations\NoRoute;
+use FOS\RestBundle\Controller\Annotations\RouteResource;
```
Comment (xabbuh) [inline Tests/Fixtures/Controller/OldTypeHintedController.php:16]: This class could also make use of the old `@Version` annotation.

Comment (xabbuh) [reply Tests/Fixtures/Controller/OldTypeHintedController.php:16]: and `@Prefix` as well

**Comment by alexander-schranz** [comment]
As @Tobion suggested not use the FOS namespace I moved the package to the *Handcrafted in the Alps* namespace and it will be published under `handcraftedinthealps/rest-routing-bundle`.

File: composer.json [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        ],
+        "exclude-from-classmap": [
+            "Resources/",
+            "Tests/"
+        ]
+    },
+    "require": {
+        "php": "^7.2",
+        "doctrine/inflector": "^1.0",
+        "friendsofsymfony/rest-bundle": "^2.8 || ^3.0",
```
Comment (Tobion) [inline composer.json:27]: Does this really require the rest-bundle? Wouldn't it be possible to make this an optional dependency?

File: Routing/Loader/Reader/RestActionReader.php [hunk 10]
```diff
@@ -0,0 +1,631 @@
+<?php
+
+/*
+ * This file is part of Handcrafted in the Alps - Rest Routing Bundle Project.
+ *
+ * (c) Sulu GmbH <hello@sulu.io>
```
Comment (Tobion) [inline Routing/Loader/Reader/RestActionReader.php:6]: Does MIT allow to just change the copyright notice?

File: composer.json (hunk 9 above)
Comment (alexander-schranz) [reply composer.json:27]: Good catch can be a dev dependency just need to conflict not matching rest bundle versions.

File: Routing/Loader/Reader/RestActionReader.php (hunk 10 above)
Comment (alexander-schranz) [reply Routing/Loader/Reader/RestActionReader.php:6]: Oh thats a good question didn't thought about that 🙈 Does somebody know?
Else if its ok for you I could just let the FOS header.

Comment (alexander-schranz) [reply Routing/Loader/Reader/RestActionReader.php:6]: As I read the copy holders need to stay and multiple can be added.

**Comment by alexander-schranz** [comment]
I merged the actual state to make it easier to test for others, its registered in packagist under `handcraftedinthealps/rest-routing-bundle`. Will make a tag as soon as we have a release of the FOSRestBundle. You still can give feedback about the code here.

**Pull request merged**